x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2397 @@
+      diff: {
+        /* [...] */
+      },
+    };
+  });
+};
+```
+
+This endpoint will need further changes, which will be detailed further down, and are out of the scope of Saved Object, migration and rule schema updates.
+
+- [**Perform Rule Upgrade** - `POST /prebuilt_rules/upgrade/_perform` (Internal)](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_upgrade/perform_rule_upgrade_route.ts)
+
+This endpoint will require major changes to add the capability of letting users selecting a custom version of the rule with custom values for rule fields. This will be explained further below in the "Changes to upgrade `_review` and `_perform` endpoints" section.
+
+The calculation of the value for the `prebuilt.isCustomized` field for the updated rule will depend on that logic as well, so its calculation will be explained in that section.
+
+However, once that value is calculated by the endpoint handler logic, we will need to pass it to the `upgradePrebuiltRules` method that is used by the handler to actually upgrade the rules. This method either patches existing rules -for the normal case-, or deletes an existing rule and recreates it if the rule underwent a type change during the upgrade (for example, the previous version of the rule had type of `query` and the new version is `eql`).
+
+The changes in the `upgradePrebuiltRules` method need to take into account both paths. In the endpoint logic handler, when calculating if the rule was customized by the user, we will create a boolean called `isRuleCustomizedDuringUpgrade`, that we will pass as an argument to `upgradePrebuiltRules`:
+
+_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_upgrade/perform_rule_upgrade_route.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_upgrade/perform_rule_upgrade_route.ts)_
+
+```ts
+// [... file continues above ...]
+
+export const performRuleUpgradeRoute = (router: SecuritySolutionPluginRouter) => {
+  router.versioned
+    .post({
+      access: 'internal',
+      path: PERFORM_RULE_UPGRADE_URL,
+      // [...]
+    })
+    .addVersion(
+      {
+        // [...]
+      },
+      async (context, request, response) => {
+        // [...]
+
+        try {
+          // [...]
+
+          // Endpoint will include logic for calculating which rule version should the current rule be upgraded to,
+          // as well as calculating the isRuleCustomizedDuringUpgrade boolean.
+          // Details explained further below in this document.
+
+          // Perform the upgrade, but pass in `isRuleCustomizedDuringUpgrade` as an argument
+          const { results: updatedRules, errors: installationErrors } = await upgradePrebuiltRules(
+            rulesClient,
+            targetRules,
+            isRuleCustomizedDuringUpgrade
+          );
```

Comment:
Shouldn't we calculate, if a rule has been customized during upgrade, per each rule separately? Why `isRuleCustomizedDuringUpgrade` is passed as a single value along with an array of `targetRules`?